src/Components/Web.JS/src/Rendering/FocusOnNavigate.ts

```diff
@@ -0,0 +1,99 @@
+
+  // Focus the element on the initial page load.
+  if (document.readyState === 'loading') {
+    // There may be streaming updates on the initial page load, so we want to allow
+    // those to add elements matching the active selector.
+    allowFocusOnEnhancedLoad = true;
+    document.addEventListener('DOMContentLoaded', afterInitialPageLoad);
+  } else {
+    afterInitialPageLoad();
+  }
+}
+
+function afterInitialPageLoad() {
+  tryApplyFocus(/* forceMoveFocus */ false);
+}
+
+function onEnhancedNavigationStart(ev: EnhancedNavigationStartEvent) {
+  // Only focus on enhanced load if the enhanced navigation is not a form post.
+  allowFocusOnEnhancedLoad = ev.options.method !== 'post';
+}
+
+function onEnhancedLoad() {
+  if (allowFocusOnEnhancedLoad) {
+    tryApplyFocus(/* forceMoveFocus */ true);
+  }
+}
+
+function onFocusIn() {
+  // As soon as an element get successfully focused, don't attempt to focus again on future
+  // enhanced page updates.
+  allowFocusOnEnhancedLoad = false;
+}
+
+let lastFocusedElement: Element | null = null;
+
+function onFocusOut(ev: FocusEvent) {
+  // It's possible that the element lost focus because it was removed from the page,
+  // and now the document doesn't have an active element.
+  // This could have happened either because an enhanced page update removed the element or
+  // because the focused element was replaced during the transition to interactivity
+  // (see https://github.com/dotnet/aspnetcore/issues/42561).
+  // In either case, we'll attempt to reapply focus only if:
+  // [1] that element was the one we last focused programmatically, and
+  // [2] it's about to get removed from the DOM.
+  const target = ev.target;
+  if (target instanceof Element && target === lastFocusedElement) {
+    // We want to apply focus after all synchronous changes to the DOM have completed,
+    // including the potential removal of this element.
+    setTimeout(() => {
+      if (!document.contains(target)) {
+        tryApplyFocus(/* forceMoveFocus */ false);
+      }
+    }, 0);
+  }
+}
+
+function tryApplyFocus(forceMoveFocus: boolean) {
+  // Don't apply focus if there's already a focused element and 'forceMoveFocus' is false.
+  // See also: https://developer.mozilla.org/docs/Web/API/Document/activeElement#value
+  if (!forceMoveFocus && document.activeElement !== null && document.activeElement !== document.body) {
+    return;
+  }
+
+  lastFocusedElement = null;
+
+  const selector = currentFocusOnNavigateElement?.getAttribute('selector');
+  if (selector) {
+    lastFocusedElement = domFunctions.focusBySelector(selector);
+  }
+}
+
+class FocusOnNavigateElement extends HTMLElement {
+  connectedCallback() {
+    // eslint-disable-next-line @typescript-eslint/no-this-alias
+    currentFocusOnNavigateElement = this;
+  }
+
+  disconnectedCallback() {
+    if (currentFocusOnNavigateElement === this) {
+      currentFocusOnNavigateElement = null;
+    }
+  }
+}
```

Comment:
In a previous iteration, I had a slightly more sophisticated implementation that gracefully handled having multiple `<FocusOnNavigate>` components at the same time, which can be later removed from the page in an arbitrary order without breaking things. However, I realized that's unlikely to be a real scenario, because `<FocusOnNavigate>` is almost always going to exist in one location (in the router), especially considering it requires a `RouteData` parameter that you can only obtain from the `<Router>` (unless you construct it yourself, which I don't think is common).

This simplified implementation always prioritizes the last-rendered `<blazor-focus-on-navigate>` custom element and assumes that if one already exists, it's going to be disconnected soon because we're in the process of applying updates to the DOM.

If others disagree, I'm happy to add the old logic back. It's really not hugely complicated and I'm confident that it works, but it's just more code that apps need to download (and that we need to maintain).

Reply:
We could choose to throw if we find more than one instance of focus on navigate on the document on the server even if we wanted. 

It's not even clear what the behavior is when many of them are available on the page interactively.

Reply:
TBH I'd be happy with avoiding the need for extra checks/errors and just go with a "first rendered wins" rule. It would be the same as the HTML `autofocus` attribute:

![image](https://github.com/user-attachments/assets/0ee35507-45b9-4959-a159-90eace4f60e6)

Reply:
That's fine for me, given that there's defined behavior. 

We should also consider interactions between `FocusOnNavigate` and `autofocus` now that you mention it.

Reply:
I've updated the implementation to defer to `autofocus` on the initial page load. That is, `autofocus` takes precedence over `FocusOnNavigate`.

That said, `autofocus` is not a replacement for `FocusOnNavigate`, because after the first element gets autofocused, no other elements can get autofocused until a full page reload. In other words, `autofocus` does not work with enhanced nav.

Reply:
Sounds reasonable.